### PR TITLE
Add kafka and zookeeper examples

### PR DESCRIPTION
**Pull request by alex-leonhardt**
Adds example deployments of Kafka and Zookeeper using StorageOS

### CODE REVIEW

**Comment by Arau** [comment]
Is there any reason for what the labels related to Helm are left? Is that on purpose, or could we remove them? 
I'd personally keep `app` and even the `chart`, but I'd remove `heritage` and `release` to ensure that there is no collisions with peoples' Helm install.

**Comment by alex-leonhardt** [comment]
In the docs I refer to these files having been rendered via Helm, so thought it'd be better _not_ to remove those labels, can do though, no problem.

**Comment by Arau** [comment]
> In the docs I refer to these files having been rendered via Helm, so thought it'd be better _not_ to remove those labels, can do though, no problem.

According to that, it makes sense to leave them.

**Comment by alex-leonhardt** [comment]
Okay well I didn't see your latest comment and removed them, I think you're right, it's better to remove them to not interfere with other things installed via Helm.

File: k8s/examples/kafka/01-rbac.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+---
+# Source: kafka/templates/rbac.yaml
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: kafka
+  namespace: default
```
Comment (Arau) [inline k8s/examples/kafka/01-rbac.yaml]: The namespace definition blocks the user to be namespace agnostic. 
Could we please remove the `namespace: default`.

File: k8s/examples/kafka/01-rbac.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: Role
+metadata:
+  name: kafka
+  namespace: default
```
Comment (Arau) [inline k8s/examples/kafka/01-rbac.yaml]: Could we please remove the namespace: default.

File: k8s/examples/kafka/01-rbac.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+metadata:
+  name: kafka
+roleRef:
+  kind: Role
+  name: kafka
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+  - kind: ServiceAccount
+    name: kafka
+    namespace: default
```
Comment (Arau) [inline k8s/examples/kafka/01-rbac.yaml]: Could we please remove the namespace: default.

File: k8s/examples/kafka/89-kafka-test-client.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kafka-test-client
+  namespace: default
```
Comment (Arau) [inline k8s/examples/kafka/89-kafka-test-client.yaml]: Could we please remove the namespace: default.

File: k8s/examples/zookeeper/89-zookeeper-client.yaml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: zookeeper-test-client
+  namespace: default
```
Comment (Arau) [inline k8s/examples/zookeeper/89-zookeeper-client.yaml]: Could we please remove the namespace: default.

**Review by Arau — COMMENTED** [review]
- Hardcoding the namespace in the manifests makes the use case less portable. 
- There is a comment at the beginning of the file coming from the Helm debug output indicating the template that the manifest is built from can lead to confusion to people that go straight to read the YAMLs without reading the public docs as those locations `templates/` don't exist in our repo. 
- Zookeper doesn't have a service account, I think it should have by default
- There are no READMEs for both ZooKeeper and Kafka. They need a README even though the public docs hold most of the information. I'd add at least a comment referencing the public docs and explaining a bit (copy&paste from the docs) what is in there. For instance it is necessary to make sure that anyone landing into the Kafka directory have an easy way to deploy zookeeper. The work is there, but people won't necessary know. It would be enough by adding a line such as `kubectl create -f ../zookeeper/`. 
The other use cases have a README that might serve as a guide, for instance the postgress one.

File: k8s/examples/kafka/01-rbac.yaml (hunk 1 above)
Comment (alex-leonhardt) [reply k8s/examples/kafka/01-rbac.yaml]: removed

File: k8s/examples/kafka/01-rbac.yaml (hunk 2 above)
Comment (alex-leonhardt) [reply k8s/examples/kafka/01-rbac.yaml]: removed

File: k8s/examples/kafka/01-rbac.yaml (hunk 3 above)
Comment (alex-leonhardt) [reply k8s/examples/kafka/01-rbac.yaml]: removed

File: k8s/examples/kafka/89-kafka-test-client.yaml (hunk 4 above)
Comment (alex-leonhardt) [reply k8s/examples/kafka/89-kafka-test-client.yaml]: removed

File: k8s/examples/zookeeper/89-zookeeper-client.yaml (hunk 5 above)
Comment (alex-leonhardt) [reply k8s/examples/zookeeper/89-zookeeper-client.yaml]: removed

**Comment by alex-leonhardt** [comment]
> * Zookeper doesn't have a service account, I think it should have by default

the YAMLs were built by rendering the helm chart, so in this case, there was no service account - I don't mind adding a service account, but then it's no longer a helm rendered manifest that someone could refer back to ? if having a service account is now standard, even if that service doesn't interact with the kubernetes api, then I can add it in, no probs, but service accounts as I understand it are only really required to interact with k8s apis (please correct me if that is wrong?)

**Comment by alex-leonhardt** [comment]
> * There are no READMEs for both ZooKeeper and Kafka. They need a README even though the public docs hold most of the information. I'd add at least a comment referencing the public docs and explaining a bit (copy&paste from the docs) what is in there. For instance it is necessary to make sure that anyone landing into the Kafka directory have an easy way to deploy zookeeper. The work is there, but people won't necessary know. It would be enough by adding a line such as `kubectl create -f ../zookeeper/`.

I don't agree with duplicating documentation, when / if the docs change, we have to try find all places we copied them to, happy to just link to the use-case doc once it's up though?

**Comment by Arau** [comment]
> > * There are no READMEs for both ZooKeeper and Kafka. They need a README even though the public docs hold most of the information. I'd add at least a comment referencing the public docs and explaining a bit (copy&paste from the docs) what is in there. For instance it is necessary to make sure that anyone landing into the Kafka directory have an easy way to deploy zookeeper. The work is there, but people won't necessary know. It would be enough by adding a line such as `kubectl create -f ../zookeeper/`.
> 
> I don't agree with duplicating documentation, when / if the docs change, we have to try find all places we copied them to, happy to just link to the use-case doc once it's up though?

I'm not suggesting to maintain 2 copies of the same docs. That as you mentioned adds a lot of maintainability hassle. However, a lot of people don't go to the official docs, even though they should. Hence they ended up deploying manifests without fully going in depth. When that doesn't work straight, we are asked to help, etc. Hence creating an avoidable overhead. 
I'd suggest to add at least a bit of information. When I refer to copy and paste I merely refer to an introduction of what is in there, not in detail (which is likely to change in the future). 
The other use case READMEs can serve as an example. They are not copies of the full doc. Just enough information (what is enough information is subject to opinion of course :-)  ).

**Comment by Arau** [comment]
> > * Zookeper doesn't have a service account, I think it should have by default
> 
> the YAMLs were built by rendering the helm chart, so in this case, there was no service account - I don't mind adding a service account, but then it's no longer a helm rendered manifest that someone could refer back to ? if having a service account is now standard, even if that service doesn't interact with the kubernetes api, then I can add it in, no probs, but service accounts as I understand it are only really required to interact with k8s apis (please correct me if that is wrong?)

I understand that, and it is not a big deal. When asking for a service account, I have in mind security issues related to PSPs or OpenShift SCCs. When I tried to deploy into OpenShift, I can't extend adding the SCC without the SA as I would have to give privileges to the default account of the project. The same will happen to any K8S cluster that enables PSP. 
The helm chart is not designed to support PSP I imagine. 

I don't think this is a big deal as any user that uses PSP, SCCs, etc can adapt the  manifests to add its own SA.
The suggestion is to follow the pattern we have been using, as all use cases have service accounts.

**Review by Arau — APPROVED** [review]
Looks great to me!

**Pull request merged**